psi4/src/psi4/libscf_solver/sad.cc

```diff
@@ -767,40 +767,73 @@ void HF::compute_SAD_guess() {
-    Dimension sad_dim(Da_->nirrep(), "SAD Dimensions");
+    if (natorb) {
+        // SAD natural orbitals (doi:10.1021/acs.jctc.8b01089)
+
+        // Number of basis functions
+        auto nbf(basisset_->nbf());
+        // Grab the density matrix
+        auto Dhelp = std::make_shared<Matrix>("Helper density matrix", AO2SO_->colspi(), AO2SO_->colspi());
+        Dhelp->copy(guess->Da());
+        // Take its negative so that most strongly occupied orbitals come first
+        Dhelp->scale(-1.0);
+
+        // Transfrom to an orthonormal basis
+        Dhelp->transform(S_);
+        Dhelp->transform(X_);
+        if (debug_) {
+            outfile->Printf("SAD Density Matrix (orthonormal basis):\n");
+            Dhelp->print();
+        }
 
-    for (int h = 0; h < Da_->nirrep(); h++) {
-        int nso = Ca_sad->rowspi()[h];
-        int nmo = Ca_sad->colspi()[h];
-        if (nmo > X_->colspi()[h]) nmo = X_->colspi()[h];
+        // Diagonalize the SAD density and form the natural orbitals
+        auto Cno_temp_ = SharedMatrix(factory_->create_matrix("SAD NO temp"));
+        Dhelp->diagonalize(Cno_temp_, epsilon_a_);
+        if (debug_) {
+            outfile->Printf("SAD Natural Orbital Occupations:\n");
+            epsilon_a_->print();
+        }
+        Ca_->gemm(false, false, 1.0, X_, Cno_temp_, 0.0);
 
-        sad_dim[h] = nmo;
+        // Same orbitals for beta
+        Cb_->copy(Ca_);
+        epsilon_b_->copy(*epsilon_a_);
 
-        if (!nso || !nmo) continue;
+    } else {
+        SharedMatrix Ca_sad = guess->Ca();
+        SharedMatrix Cb_sad = guess->Cb();
+        Da_->copy(guess->Da());
+        Db_->copy(guess->Db());
+        Dimension sad_dim(Da_->nirrep(), "SAD Dimensions");
 
-        double** Cap = Ca_->pointer(h);
-        double** Cbp = Cb_->pointer(h);
-        double** Ca2p = Ca_sad->pointer(h);
-        double** Cb2p = Cb_sad->pointer(h);
+        for (int h = 0; h < Da_->nirrep(); h++) {
+            int nso = Ca_sad->rowspi()[h];
+            int nmo = Ca_sad->colspi()[h];
+            if (nmo > X_->colspi()[h]) nmo = X_->colspi()[h];
 
-        for (int i = 0; i < nso; i++) {
-            ::memcpy((void*)Cap[i], (void*)Ca2p[i], nmo * sizeof(double));
-            ::memcpy((void*)Cbp[i], (void*)Cb2p[i], nmo * sizeof(double));
-        }
-    }
+            sad_dim[h] = nmo;
 
-    nalphapi_ = sad_dim;
-    nbetapi_ = sad_dim;
-    nalpha_ = sad_dim.sum();
-    nbeta_ = sad_dim.sum();
-    doccpi_ = sad_dim;
-    soccpi_ = Dimension(Da_->nirrep(), "SAD SOCC dim (0's)");
+            if (!nso || !nmo) continue;
 
-    energies_["Total Energy"] = 0.0;  // This is the -1th iteration
+            double** Cap = Ca_->pointer(h);
+            double** Cbp = Cb_->pointer(h);
+            double** Ca2p = Ca_sad->pointer(h);
+            double** Cb2p = Cb_sad->pointer(h);
+
+            for (int i = 0; i < nso; i++) {
+                ::memcpy((void*)Cap[i], (void*)Ca2p[i], nmo * sizeof(double));
```

Comment:
Can we switch this to `std::copy` while we are here?

Reply:
These are not `std::vector`!

Reply:
Also, are the arguments the wrong way around? Or is this due to some weird C-Fortran layout difference.

Reply:
Right, not enough coffee... `C_DCOPY(Cap[I], 1, Ca2p, 1, nmo);`? We seem to be moving away from raw `memcpy` operations to prevent bytes vs elements confusion.

Reply:
Have some more coffee, you had the size argument in the wrong place ;)

Actually, it seems that with the argument in the beginning the SAD energies are screwed up, so there's some other issue with `C_DCOPY`...

Reply:
Well... thats very odd. With more coffee your version looks correct. I will apply the @andysim iterative coffee technique.

Reply:
@dgasmith the arguments were the wrong way around as well :D

Reply:
Ah, BLAS...